File: Leetcode/easy/QueueUsingStacks.scala
// Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).
// Amortized O(1) runtime

import scala.collection.mutable.Stack

class MyQueue() {
    val inStack = Stack[Int]()
    val outStack = Stack[Int]()

    def push(x: Int) = {
        inStack.push(x)
    }

    def pop(): Int = {
        peek()
        outStack.pop
    }

    def peek(): Int = {
        if (outStack.isEmpty) {
            while (!inStack.isEmpty) {
                outStack.push(inStack.pop)
            }
        }
        outStack.top
    }

    def empty(): Boolean = {
        inStack.isEmpty && outStack.isEmpty
    }
}

File: Leetcode/easy/QueueUsingStacks.js
// Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).
// Amortized O(1) runtime

const MyQueue = function () {
    this.input = [];
    this.output = [];
};

MyQueue.prototype.push = function (x) {
    this.input.push(x);
};

MyQueue.prototype.pop = function () {
    this.peek();
    return this.output.pop();
};

MyQueue.prototype.peek = function () {
    if (this.output.length === 0) {
        while (this.input.length !== 0) {
            this.output.push(this.input.pop());
        }
    }
    return this.output[this.output.length - 1];
};

MyQueue.prototype.empty = function () {
    return this.input.length === 0 && this.output.length === 0;
};